### ISSUE TITLE
Improving save_on_cpu's performance by overlapping memory transfers with compute

### ISSUE THREAD

**Issue by hanzhi713** [pytorch/pytorch#106360]
### 🚀 The feature, motivation and pitch

The current implementation invokes h2d/d2h on the same stream as compute, essentially blocking the GPU from doing other computation. A straightforward approach to improve the performance is to call h2d/d2h on a different stream, allowing GPU's DMA engine work concurrently with other computation. 

I got a prototype working, but my limited knowledge of pytorch internals prevented my from making this perfect. So I would like to get some help and suggestions, and hopefully making this into pytorch. My code is shown below and the points for discussion are written in the comments.

```python3
from torch.autograd.graph import saved_tensors_hooks
import torch
from typing import Tuple
from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import ActivationWrapper
from torch.distributed.fsdp._limiter_utils import _FreeEventQueue


class faster_save_on_cpu(saved_tensors_hooks):
    copy = None
    queue = _FreeEventQueue()

    def __init__(self):
        if faster_save_on_cpu.copy is None:
            # make sure we only create the h2d/d2h stream once
            faster_save_on_cpu.copy = torch.cuda.Stream()

        def pack_to_cpu(tensor: torch.Tensor):
            # Here, we may redundantly save full weights tensors to cpu. 
            # For example, by default, a linear layer will save its input, weight 
            # and bias for backward. However, if we already have FSDP, we probably 
            # don't want save the **unsharded** weight and bias to CPU and want to 
            # leave them on GPU to reduce memory transfers. Is there a way to skip 
            # saving weight tensors?

            assert not tensor.is_sparse  # not yet supported
            if not tensor.is_cuda:
                return None, tensor

            self.limit_prefetch()
            faster_save_on_cpu.copy.wait_stream(torch.cuda.current_stream())
            with torch.cuda.stream(faster_save_on_cpu.copy):
                cpu_tensor = tensor.to("cpu", non_blocking=True)
            self.record_event()
            tensor.record_stream(faster_save_on_cpu.copy)
            return (tensor.device, cpu_tensor)

        def unpack_from_cpu(packed: Tuple[torch.device, torch.Tensor]):
            device, cpu_tensor = packed
            if device is None:
                return cpu_tensor

            self.limit_prefetch()
            with torch.cuda.stream(faster_save_on_cpu.copy):
                # here, we must allocate tensor on the copy stream to allow
                # copy to start early. If pytorch can provide a method to
                # allocate an empty block **immediately** rather than reusing a
                # previous block on the compute stream, the problem below can be solved. 
                tensor = cpu_tensor.to(device, non_blocking=True)
            self.record_event()

            # subsequent gradient computation need to wait for h2d to complete
            torch.cuda.current_stream().wait_stream(faster_save_on_cpu.copy)

            # Now we have a problem. Ideally, we want to avoid this memcpyD2D to save time and memory, 
            # but the tensor is allocated on a different stream as opposed to the compute stream, so
            # to make this work, we need to call .record_stream after this tensor's lifetime ends
            # during backward. However, autogard seems like a blackbox to me, making it not 
            # practical to insert a .record_stream(). As a workaround, we can clone this tensor. 

            # I've tried hacking the destructor of out_tensor by subclassing and overriding Tensor's __del__,
            # but this introduced significant python overhead due to all subsequent tensors having to call
            # .record_stream(). 
            out_tensor = tensor.clone()
            tensor.record_stream(torch.cuda.current_stream())
            return out_tensor

        super().__init__(pack_to_cpu, unpack_from_cpu)

    def limit_prefetch(self):
        # limit the number of h2d/d2h submitted to reduce peak memory usage, like FSDP's limit_all_gathers option
        prev_ev = self.queue.dequeue_if_needed()
        if prev_ev:
            prev_ev.synchronize()

    def record_event(self):
        event = torch.cuda.Event()
        event.record(faster_save_on_cpu.copy)
        self.queue.enqueue(event)

```

**Comment by ezyang** [comment]
You basically need AsyncCpuTensor, which we've talked about for years, but because overhead matters a lot you can't just do the easy way out of using a Python subclass.

Today, it is probably more promising to work out some hybrid strategy with PT2 to remove overhead. cc @bdhirsh as another potential tensor subclass PT2 use case.

**Comment by ezyang** [comment]
I am fairly confused; if you just wanted to move data from one gpu to another, why not just do direct P2P communication

**Comment by hanzhi713** [comment]
> I am fairly confused; if you just wanted to move data from one gpu to another, why not just do direct P2P communication

It's not about moving between GPUs. It's about moving between CPU and GPU. The whole idea of `save_on_cpu` is that during forward pass, we offload activations to CPU memory to reduce peak GPU memory usage. During backward pass, we move the CPU activations back to GPU on-demand for computing gradients.

The current implementation of `save_on_cpu` in pytorch does this in the same stream as computation, basically blocking GPUs from doing other stuff. I proposed doing these CPU <-> GPU transfers in a different stream so that GPU can perform computation at the same time, but I'm having some troubles to make this perfect.

**Comment by hanzhi713** [comment]
My current code snippet already made sure that HtD will happen after all outstanding DtH finishes, because they are stream ordered. Additionally, this statement `torch.cuda.current_stream().wait_stream(faster_save_on_cpu.copy)` will ensure that gradient computation will not begin before activation has finished moving to the GPU. 

So the problem is not related to the CPU tensor in anyway. The problem is related to the GPU tensor allocated in the copy stream. When we do HtD in the `unpack_from_cpu ` hook, we must first allocate an empty GPU tensor. This occurred implicitly in `tensor = cpu_tensor.to(device, non_blocking=True)`. However, this **GPU** tensor is allocated in the `copy` stream, so when its lifetime goes out in the autograd computation, it might be reused **before** the gradient computation kernels finish by another allocation in the `copy` stream.

**Comment by zbtrs** [comment]
Is the problem solved?

**Comment by awgu** [comment]
I think @janeyx99 is working on something for this.

If you want some ideas, you can also look at https://github.com/pytorch/torchtitan/pull/467.

**Comment by janeyx99** [comment]
@zbtrs I've written offloading with streams in torchtune here (heuristically tuned for llama w/ activations checkpointing): https://github.com/pytorch/torchtune/blob/main/torchtune/training/_activation_offloading.py. Feel free to try it out as a context manager for your model fwd + loss calculation, or as inspiration for working with streams (which, I agree, are not the most straightforward). The streams impl requires being on torch 2.5 or later as we make use of Autograd node post hooks.
